### DISCUSSION TITLE
Add more data structures to `package collection`

### DISCUSSION THREAD

**Discussion by gingerBill** [odin-lang/Odin#977] [Proposals]
More data structures to add:

* [x] Array
* [x] Queue
* [x] Map (Hash Table)
* [x] Small Array
* [x] Set
* [x] Bit Array
* [ ] B Tree
* [ ] B+ Tree
* [ ] R Tree
* [ ] Red-Black tree
* [ ] Range Tree
* [ ] Heap
* [ ] Fibonacci Heap
* [ ] Ring
* [ ] Skip list
* [ ] C Trie
* [ ] D Trie

**Comment by kevinw** [comment]
@gingerBill I thought Small Array would be something more like https://github.com/KhronosGroup/SPIRV-Cross/blob/master/spirv_cross_containers.hpp#L191 -- where the backing store begins on the stack, but is automatically moved to the heap for you if the number of elements grows too big. This extra bit of functionality seems to be missing from your Small Array. We already have an allocator that can point at stack allocated memory.

**Comment by gingerBill** [comment]
@kevinw It might be better to call it a `Fixed_Small_Array` and then have your `Dynamic_Small_Array` be a separate thing you'd like.

**Comment by kevinw** [comment]
Yeah, that would make sense! Although I wonder about the need for a Fixed_Small_Array, really...

**Comment by gingerBill** [comment]
I use the construct of a "fixed small array" all the time, especially when I want a small inline data structure that requires no external allocator.

**Comment by ghost** [comment]
What about also adding a Ternary Search Tree. I think I seen these would be useful for spellcheck type of things.

**Comment by ghost** [comment]
Dense and Sparse Spatial Hashing

**Comment by gingerBill** [comment]
This entire package needs a redesign and rethinking #782

**Comment by saccarosium** [comment]
Hi, how up to date is the checklist?

I would like to try implement the `Heap` myself.

Also I was thinking that maybe introduce a `Stack` type that simply uses the `Queue`. I find beneficial distinguish between the two for clarity, while the behaviour is the same if I just used the `Queue` and than do just `push_back` and `pop_back`, when I read the code I find myself having to keep in mind: "I'm using this queue as a stack" witch I don't really like if I can avoid.
